int_vector library

havogt opened the pull request:
Adds the int_vector library (fixes #1669).

for manipulation of offsets, sizes and extents in `fn`.

`int_vector` is a `hymap` which value types are all either of integral types or (gridtools) integral constant types with the following functions defined
- `plus`  -- a function that takes two `int_vector`'s and return an `int_vector`. The dimensions (keys) of the result is the the union of the arguments keys. 
- `multiply` -- a function that accepts an `int_vector` and a scalar (integral or integral_constant) and returns an `int_vector`
- `prune_zeros`  -- a function that takes an`int_vector` and removes dimensions which values are equal to `0_c` 

Additionally, arithmetic operators are defined in namespace `int_vector::arithmetic`.

See #1669 for more details.

anstaf reviewed (commented):
Let us
 * assume that we only deal with `gridtools::integer_constant`'s 
 * implement host/device functionality in a simple way

(see my comments on #1669 for more details)

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+        };
+
+        template <class T0, class T1>
+        struct value_t_merger<T0, T1> {
```
my version is:
```c++

template <class...> struct value_t_merger;

template <template <class...> class Item, class Dir, class... Vs>
struct value_t_merger<Item<Dir, Vs>...> {
   using type = Item<Dim, std::decay_t<decltype((Vs{} + ...))>>;
};
```

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+        template <class... Ts>
+        struct value_t_merger;
+
+        template <class T>
```
this will never be instantiated

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+                    template <class First, class Second>
+                    GT_TARGET GT_FORCE_INLINE GT_TARGET_CONSTEXPR Type operator()(
+                        First &&first, Second &&second) const {
+                        return at_key_with_default<Key, std::integral_constant<Type, 0>>(wstd::forward<First>(first)) +
```
how about:
```c++
template <class L, class R>
GT_FUNCTION constexpr auto operator()(L const& l, R const& r) const {
   if constexpr (not has_key<L, Key>())
     return host_device::at_key<Key>(r);
   if constexpr (not has_key<R, Key>())
     return host_device::at_key<Key>(l);
   return host_device::at_key<Key>(l) + host_device::at_key<Key>(r); 
} 
```
Note that there are no forwards here -- args can not be rvalues because they passed to the several generators

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+                }
+
+                template <class Vec, class Scalar>
+                GT_TARGET GT_FORCE_INLINE GT_TARGET_CONSTEXPR auto multiply(Vec &&vec, Scalar &&scalar) {
```
Pass scalar by value

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+
+                template <class Vec, class Scalar>
+                GT_TARGET GT_FORCE_INLINE GT_TARGET_CONSTEXPR auto multiply(Vec &&vec, Scalar &&scalar) {
+                    return tuple_util::transform([&scalar](auto &&v) { return v * scalar; }, std::forward<Vec>(vec));
```
`tuple_util::transform` is an alias of `tuple_util::host::transform`. Is is not what you want here

anstaf replied:
`[scalar](auto v) { return v * scalar; }` is better and simpler

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+            using int_vector_detail::plus;
+
+            template <class First, class... Rest>
+            GT_TARGET GT_FORCE_INLINE GT_TARGET_CONSTEXPR auto plus(First && first, Rest && ...rest) {
```
I would either skip this or implement variadic version directly

anstaf commented on tests/unit_tests/common/test_int_vector.cpp:
```diff
@@ -0,0 +1,78 @@
+namespace gridtools {
+    namespace {
+        using namespace literals;
+
```
please test that the stuff works with `gridtools::tuple` and `gridtools::array`

anstaf commented on tests/unit_tests/common/test_int_vector.cpp:
```diff
@@ -0,0 +1,78 @@
+    namespace {
+        using namespace literals;
+
+        struct a;
```
We need device tests probably

havogt commented:
Thanks for the review. I implemented the changes and added the missing pieces.
I didn't implement `operator/` as this cannot be implemented straight-forward in terms of multiply (because 1/v == 0) and probably there is also no use-case?!

havogt commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+
+            template <class Vec, std::enable_if_t<is_int_vector_v<Vec>, bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator-(Vec &&vec) {
+                return multiply(std::forward<Vec>(vec), integral_constant<int, -1>{}); // TODO not compiling with nvcc
```
I didn't try to investigate why the test for negate doesn't compile with nvcc (note it's only host compilation). Maybe we can have a quick look together as you have much better intuition for nvcc behavior.

anstaf replied:
I am happy to help, ping me any time

havogt replied:
Not sure if it's worth debugging. It triggers a compile error in the test in
```
static_assert(std::is_same_v<integral_constant<int, -2>, std::decay_t<decltype(at_key<c>(testee))>>);
```
However if I look at the type outside of the static_assert, all seems fine.

anstaf replied:
No, no. Let us debug it.

havogt commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,134 @@
+    inline constexpr bool is_int_vector_v = is_int_vector<T>::value;
+
+    namespace int_vector {
+        using int_vector_impl_::multiply;
```
I think you sometimes use this style to hide some complex implementation in the impl namespace. What's your rationale when to do it?

anstaf replied:
Good question. Let me cover this topic here. Probably we should revive gridtools C++ style guide in some form an place the rationales there (applicable mostly for a new code, old code might violate):
- The directory structure of the `include` should be in sync with c++ namespaces. 
  - for any namespace `n` should be either a directory `n` or a header `n.cpp`
  - for any directory `dir` (that is not `common`), all c++ entities that are defined in that directory should be within the `dir` namespace. 
  - for `dir/common` directory all c++ entities are defined within the `dir` namespace
  - if directory has public subdirectories (`internal`, `detail`, `private`, `impl` etc. are considered not public), it should not have headers directly. The exceptions are umbrella headers (like `gridtools/meta.hpp`) which contains only `#include` directives pointed to the headers within the subdirectory  of the same name. This is kind of auxiliary recommendation. It allows to keep explorer pane in IDEs neat and compact.   
  - if some directory do not follow those rules for the reason, it is better to have `README.md` in it. Like in `gridtools/stencil`.
- headers should place all private c++ entities into implementation namespace. This namespace should be unique (two headers should not share the same implementation namespace). The usual pattern is to use `name::impl_` if the header   `name.hpp` manages its own namespace otherwise `name_impl_`. There are two patterns to deal with implementation namespaces:
  - define everything within the `impl_` and expose the public entities  to the outer namespace with `using`;
  - define public entities it the outer namespace and use internal stuff explicitly (`impl_::impl_foo` within `foo`).
 In some cases the second pattern doesn't because your design relies on ADL. If it is not the case, the choice should be driven by readability.
 Examples:
   - You have ten public entities in your header, only one public entity uses a private thing that is one liner -- go for the second pattern.
   - You have ten private entities and a single public one that directly uses all private things -- go for the first pattern.

fthaler commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+         * @brief Returns `int_vector` with elements removed that are `integral_constant<T, 0>`
+         */
+        template <class Vec>
+        GT_FUNCTION GT_CONSTEXPR auto normalize(Vec &&vec) {
```
I don’t like that function name, see https://github.com/GridTools/gridtools/issues/1669#issuecomment-985400152

anstaf replied:
LGTM for any name

havogt replied:
Makes sense. I'll go with the very explicit `prune_zeros` or do you have a better name?

anstaf commented:
> Thanks for the review. I implemented the changes and added the missing pieces. I didn't implement `operator/` as this cannot be implemented straight-forward in terms of multiply (because 1/v == 0) and probably there is also no use-case?!

You are right about `operator/`. Either don't implement it or implement additional `divide` (which is also one liner with a single tuple_util::host_device::transform call

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+
+#include "../meta.hpp"
+#include "defs.hpp"
+#include "gridtools/meta/debug.hpp"
```
not used

havogt replied:
Thanks for checking the includes! clangd auto includes this everytime I complete GT_META_PRINT_TYPE...

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+#include <type_traits>
+
+#include "../meta.hpp"
+#include "defs.hpp"
```
not sure it is used

havogt replied:
GT_FUNCTION is coming from there.

havogt replied:
Ah no, but GT_CONSTEXPR. Which reminds me of the question that I had yesterday: is GT_CONSTEXPR still a thing? I.e. does nvcc still not understand `constexpr` for the version we are supporting?

anstaf replied:
`nvcc` did `constexpr` wrong in several ways in several versions. 
There were two aspects that I remember:
  - sometime run time execution of `constexpr` in device degrade performance.
  - sometime `nvcc` has compilation errors. mostly when the `constexpr` functions are part of expressions in non evaluation  context (`decltype`).

Hence to check this again with CUDA 11 you need full test suite (unit tests + perfs)

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+#include "host_device.hpp"
+#include "hymap.hpp"
+#include "integral_constant.hpp"
+#include "tuple.hpp"
```
not used

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+
+namespace gridtools {
+
+    namespace int_vector_impl_ {
```
I would prefer `int_vector::impl_` as an internal namespace

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+         *
+         * The keys of the resulting `int_vector` are the union of the keys of the operands.
+         */
+        template <class First, class Second>
```
as soon as `merger_t` and `add_f` are variadic, there is no reason to keep `plus` two arguments only

havogt replied:
done

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+            using type = Item<Key, std::decay_t<decltype((Vs{} + ...))>>;
+        };
+
+        template <class... T>
```
please follow common conversion for variadic parameters (use `Ts` instead of `T`)

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+            using keys_t = meta::transform<meta::first, merged_meta_map_t>;
+            using generators = meta::transform<add_f, keys_t>;
+            return tuple_util::host_device::generate<generators, hymap::from_meta_map<merged_meta_map_t>>(
+                std::forward<First>(first), std::forward<Second>(second));
```
`wstd::forward` here and below 
+ add the appropriate header

havogt replied:
done

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+        struct is_integral_zero<I,
+            std::enable_if_t<is_integral_constant<meta::second<I>>::value && meta::second<I>::value == 0>>
+            : std::true_type {};
+
```
my version:
```c++
template <class I>
using is_constant_zero = integral_constant_of<meta::second<I>, 0>;
```

havogt replied:
done

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+    }; // TODO proper type trait: is_hymap and (is_integral or is_gr_integral_constant)
+
+    template <class T>
+    inline constexpr bool is_int_vector_v = is_int_vector<T>::value;
```
`inline` is redundant here. also it is C++20.

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+        namespace arithmetic {
+            template <class Vec,
+                class Scalar,
+                std::enable_if_t<is_int_vector_v<Vec> &&
```
here and below: `is_int_vector_v<Vec>` -> `is_int_vector_v<std::decay_t<Vec>>`

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,161 @@
+            }
+
+            template <class Vec, std::enable_if_t<is_int_vector_v<Vec>, bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator-(Vec &&vec) {
```
probably unary `operator+` should be here for  consistency :)

havogt replied:
done

anstaf commented on tests/unit_tests/common/test_int_vector.cpp:
```diff
@@ -0,0 +1,138 @@
+
+#include "gridtools/common/array.hpp"
+#include "gridtools/meta/debug.hpp"
+#include <gridtools/common/int_vector.hpp>
```
I would prefer to put a testee as a first header

anstaf commented on tests/unit_tests/common/test_int_vector.cpp:
```diff
@@ -0,0 +1,138 @@
+ */
+
+#include "gridtools/common/array.hpp"
+#include "gridtools/meta/debug.hpp"
```
not used

anstaf commented on tests/unit_tests/common/test_int_vector.cpp:
```diff
@@ -0,0 +1,138 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "gridtools/common/array.hpp"
```
use `<>`

anstaf commented on tests/unit_tests/common/test_int_vector.cu:
```diff
@@ -0,0 +1,87 @@
+
+#include "gridtools/common/array.hpp"
+#include "gridtools/meta/debug.hpp"
+#include <gridtools/common/int_vector.hpp>
```
please double check you includes, to ensure you include what you use

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,156 @@
+
+namespace gridtools {
+
+    namespace int_vector::impl_ {
```
this is c++20 syntax, do you want to keep it c++17?

havogt replied:
it is c++17

havogt replied:
But I forgot that actually our default is still c++14.

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,156 @@
+ * A type `T` is `int_vector` if
+ *  - `is_hymap<T>`
+ *  - all value types `V` are std::is_integral<V> or is_gr_integral_constant<V>
+ *  - `std::is_trivially_copy_constructible<T>` // TODO isn't that trivially satisfied by the previous conditions?
```
we do it to test `is_trivially_copy_constructible` because hymap can be implemented differently and also 
`is_gr_integral_constant` assumes that the class is derived from `gridtools::integral_constant`, not instantiated.

havogt replied:
I see

havogt commented:
launch jenkins

havogt commented:
launch jenkins

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,179 @@
+
+        template <class T>
+        using elements_are_integral_or_gr_integral_constant =
+            meta::all<meta::transform<int_vector::impl_::is_integral_or_gr_integral_constant,
```
Let us simplify it to:
```c++
template <class T>
using  elements_are_integral_or_gr_integral_constant =
    meta::all_of<is_integral_or_gr_integral_constant, tuple_util::traits::to_types<T>>;
```

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,179 @@
+    } // namespace int_vector::impl_
+
+    template <class T, class = void>
+    struct is_int_vector : std::false_type {};
```
I would prefer to have `is_int_vector` within `gridtools::int_vector` namespace and export it to the outer `gridtools` via `using`. The same is for `is_int_vector_v`.

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,179 @@
+            template <class Vec,
+                class Scalar,
+                std::enable_if_t<is_int_vector_v<std::decay_t<Vec>> &&
+                                     (std::is_integral_v<Scalar> || is_integral_constant<Scalar>::value),
```
use is_integral_or_gr_integral_constant

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,179 @@
+            template <class Vec,
+                class Scalar,
+                std::enable_if_t<is_int_vector_v<std::decay_t<Vec>> &&
+                                     (std::is_integral_v<Scalar> || is_integral_constant<Scalar>::value),
```
same

anstaf commented on include/gridtools/common/int_vector.hpp:
```diff
@@ -0,0 +1,179 @@
+
+            template <class Vec, std::enable_if_t<is_int_vector_v<std::decay_t<Vec>>, bool> = true>
+            GT_FUNCTION GT_CONSTEXPR auto operator-(Vec &&vec) {
+                return multiply(wstd::forward<Vec>(vec), integral_constant<int, -1>{}); // TODO not compiling with nvcc
```
let us try to resolve TODO. Have you tried to replace `{}` by `()`?

havogt replied:
I forgot to delete the TODO. It is working in this version

havogt commented:
Since there are fold expression in the variadic plus, I'd propose we drop C++14 support before we merge this. I'll open a PR to try if we have compatible compilers on all machines.

havogt commented:
launch jenkins

havogt commented:
launch jenkins
